### PR TITLE
Add Api key expiration checking routine

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

This PR adds a new quantum routine to check each day if the api key is expired.
When the api key expiration is less then 30 days, an email is sent.
When the api key expires an email is sent.

The email are sent until the api key is generated again with a new expiration.
No email are sent if the api key expiration is infinite

image
image

## How was this tested?

Automated tests

### CODE REVIEW

File: lib/trento/infrastructure/alerting/alerting.ex [hunk 1]
```diff
@@ -74,6 +79,34 @@ defmodule Trento.Infrastructure.Alerting.Alerting do
...
 
+  defp maybe_notify_api_key_expiration(false), do: :ok
+
+  defp maybe_notify_api_key_expiration(true) do
+    case Settings.get_api_key_settings() do
+      {:ok, %ApiKeySettings{expire_at: nil}} ->
+        :ok
+
+      {:ok, %ApiKeySettings{} = api_key_settings} ->
+        maybe_send_api_key_notification(api_key_expiration_days(api_key_settings))
```
Comment (arbulu89) [inline lib/trento/infrastructure/alerting/alerting.ex:90]:
```
api_key_settings
|> api_key_expiration_days()
|> maybe_send_api_key_notification()
```

File: lib/trento/infrastructure/alerting/alerting.ex [hunk 2]
```diff
@@ -74,6 +79,34 @@ defmodule Trento.Infrastructure.Alerting.Alerting do
...
+    deliver_notification(EmailAlert.api_key_expired())
+  end
+
+  defp maybe_send_api_key_notification(days) when days < 30 do
+    deliver_notification(EmailAlert.api_key_will_expire(days))
+  end
+
+  defp maybe_send_api_key_notification(_), do: :ok
+
+  defp api_key_expiration_days(%ApiKeySettings{expire_at: expire_at}),
```
Comment (arbulu89) [inline lib/trento/infrastructure/alerting/alerting.ex:107]: This should be moved above `maybe_send_api_key_notification` as it is called before

File: lib/trento/infrastructure/alerting/alerting.ex [hunk 3]
```diff
@@ -74,6 +79,34 @@ defmodule Trento.Infrastructure.Alerting.Alerting do
...
+        error
+    end
+  end
+
+  defp maybe_send_api_key_notification(days) when days < 0 do
+    deliver_notification(EmailAlert.api_key_expired())
+  end
+
+  defp maybe_send_api_key_notification(days) when days < 30 do
+    deliver_notification(EmailAlert.api_key_will_expire(days))
```
Comment (arbulu89) [inline lib/trento/infrastructure/alerting/alerting.ex:102]:
```
days
|> EmailAlert.api_key_will_expire()
|> deliver_notification()
```

File: lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex [hunk 4]
```diff
@@ -0,0 +1,426 @@
+<!doctype html>
+<html>
+
+<head>
+  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
+  <title>Simple Transactional Email</title>
+  <style>
```
Comment (arbulu89) [inline lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex:8]: We need to study if this styles can be imported from other file, as we use them in the other template as well.
Could we open a tech debt ticket?

File: lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+            <tr>
+              <td class="wrapper">
+                <table role="presentation" border="0" cellpadding="0" cellspacing="0">
+                  <tr>
+                    <td>
+                      <p>Hi there,</p>
+                      <p></p>
+                      <div>
+                        <%= if @api_key_expired do %>
+                          <b> Your api key is expired </b>
```
Comment (arbulu89) [inline lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex:380]: Let's add a `.` at the end.

File: lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+                        <% else %>
+                          <b> Your api key is going to expire in <%= @expire_days %> days.
+                        <% end %>
+                       </p>
+                      </div>
+                      <p>Use the Settings page in the Trento Console to generate a new api key</p>
```
Comment (arbulu89) [inline lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex:386]: Let's add a `.` at the end of the phrase

**Review by arbulu89 — APPROVED** [review]
Great job!
Just some nitpick comments

File: lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex (hunk 4 above)
Comment (CDimonaco) [reply lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex:8]: Yes, they are standard eex file, we can template and have components to import, I will open a tech debt ticket

**Review by abravosuse — APPROVED** [review]
Code apart, it looks good to me @CDimonaco !

File: lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex (hunk 5 above)
Comment (CDimonaco) [reply lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex]: done

File: lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex (hunk 6 above)
Comment (CDimonaco) [reply lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex]: done

File: lib/trento/infrastructure/alerting/alerting.ex (hunk 2 above)
Comment (CDimonaco) [reply lib/trento/infrastructure/alerting/alerting.ex]: done

File: lib/trento/infrastructure/alerting/alerting.ex (hunk 1 above)
Comment (CDimonaco) [reply lib/trento/infrastructure/alerting/alerting.ex]: done

File: lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex (hunk 4 above)
Comment (CDimonaco) [reply lib/trento/infrastructure/alerting/emails/templates/api_key_expiration.html.eex:8]: Created

**Review by rtorrero — COMMENTED** [review]
LGTM (wasn't able to review in time but still all looks good)

**Pull request merged**